src/main/java/co/cask/gcp/bigquery/sink/AbstractBigQuerySink.java

```diff
@@ -0,0 +1,419 @@
+        throw new IllegalStateException(
+          String.format("Field '%s' is of unsupported type '%s'", name, schema.getType()));
+    }
+  }
+
+  /**
+   * Child classes must provide configuration based on {@link AbstractBigQuerySinkConfig}.
+   *
+   * @return config instance
+   */
+  protected abstract AbstractBigQuerySinkConfig getConfig();
+
+  /**
+   * Child classes must override this method to provide specific validation logic to executed before
+   * actual {@link #prepareRun(BatchSinkContext)} method execution.
+   * For example, Batch Sink plugin can validate schema right away,
+   * Batch Multi Sink does not have information at this point to do the validation.
+   *
+   * @param context batch sink context
+   */
+  protected abstract void prepareRunValidation(BatchSinkContext context);
+
+  /**
+   * Executes main prepare run logic, i.e. prepares output for given table (for Batch Sink plugin)
+   * or for a number of tables (for Batch Multi Sink plugin).
+   *
+   * @param context batch sink context
+   * @param bucket bucket name
+   */
+  protected abstract void prepareRunInternal(BatchSinkContext context, String bucket) throws IOException;
+
+  /**
+   * Returns output format provider instance specific to the child classes that extend this class.
+   *
+   * @param configuration Hadoop configuration
+   * @param tableName table name
+   * @param tableSchema table schema
+   * @return output format provider
+   */
+  protected abstract OutputFormatProvider getOutputFormatProvider(Configuration configuration,
+                                                                  String tableName,
+                                                                  Schema tableSchema);
+
+  /**
+   * If provided dataset name does not exists in Big Query, creates it.
+   * This operation only takes place, if pipeline does not run in preview mode.
+   *
+   * @param previewEnabled indicates whether the pipeline is running in preview.
+   */
+  private void createDataset(boolean previewEnabled) throws IOException {
+    if (previewEnabled) {
+      BigQuery bigquery = BigQueryUtil.getBigQuery(getConfig().getServiceAccountFilePath(), getConfig().getProject());
+      // create dataset if it does not exist
+      if (bigquery.getDataset(getConfig().getDataset()) == null) {
+        try {
+          bigquery.create(DatasetInfo.newBuilder(getConfig().getDataset()).build());
+        } catch (BigQueryException e) {
+          throw new IllegalStateException("Exception occurred while creating dataset: " + getConfig().getDataset(), e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Generates full path to temporary bucket based on given bucket and table names.
+   *
+   * @param bucket bucket name
+   * @param tableName table name
+   * @return full path to temporary bucket
+   */
+  private String getTemporaryGcsPath(String bucket, String tableName) {
+    return String.format(gcsPathFormat + "/hadoop/input/%s-%s", bucket, tableName, uuid);
```

Comment:
3 params given but only two %s?

Reply:
Also, why do we need ``hadoop`` in the path? can we remove it?